compiler/src/iree/compiler/GlobalOptimization/QuantizedConvToConv.cpp

```diff
@@ -163,81 +172,206 @@ struct QuantizedConvToConv
+                                         filterFDimIndex, inputCDimIndex);
     }
 
     if (!fZpIsZero) {
-      // Reduce along the input feature dimension:
-      //   sum(a, b, c) = input(a, b, c, d)
-      Value inputSum =
-          sumReduceDimensionSubset(builder, input, accETy,
-                                   /*reduce_dim*/ {false, false, false, true});
+      newConv = applyInputZeroPointCorrection(
+          isNHWC, builder, newConv, input, filter, fZp, accETy, strides,
+          dilations, inputNDimIndex, inputHDimIndex, inputWDimIndex,
+          inputCDimIndex, filterHDimIndex, filterWDimIndex);
+    }
 
-      // Materialize a length-1 dimension at the end of the summation.
-      SmallVector<ReassociationExprs> reassociationMap(3);
-      for (int i = 0; i < 3; i++)
-        reassociationMap[i].push_back(builder.getAffineDimExpr(i));
-      reassociationMap.back().push_back(builder.getAffineDimExpr(3));
+    // Apply the final update that occurs when there are multiple zero-points.
+    if (!iZpIsZero && !fZpIsZero) {
+      newConv = applyFinalZeroPointUpdate(builder, newConv, filter, iZp, fZp,
+                                          accETy, filterHDimIndex,
+                                          filterWDimIndex, filterCDimIndex);
+    }
 
-      auto expandTy =
-          RankedTensorType::get({inputTy.getDimSize(0), inputTy.getDimSize(1),
-                                 inputTy.getDimSize(2), 1},
-                                accETy);
-      inputSum = builder.create<tensor::ExpandShapeOp>(expandTy, inputSum,
-                                                       reassociationMap);
+    rewriter.replaceOp(op, newConv);
+    return success();
+  }
 
-      llvm::SmallVector<int64_t> poolDims;
-      llvm::SmallVector<Value> poolDynDims;
-      GetDynamicDym(builder, poolDims, poolDynDims, inputSum, 0);
-      GetDynamicDym(builder, poolDims, poolDynDims, newConv, 1);
-      GetDynamicDym(builder, poolDims, poolDynDims, newConv, 2);
-      GetDynamicDym(builder, poolDims, poolDynDims, inputSum, 3);
+private:
+  bool isConvNHWC(Conv2DQOpType op) const {
+    if (auto convOp = dyn_cast<linalg::Conv2DNhwcHwcfQOp>(&op)) {
+      return true;
+    } else if (auto convOp = dyn_cast<linalg::Conv2DNchwFchwQOp>(&op)) {
+      return false;
+    } else {
+      llvm_unreachable("Unsupported Conv2DOp type!");
+    }
+  }
 
-      // Perform a sum-pooling operation across the kernel width and height.
-      auto poolTy = RankedTensorType::get(poolDims, accETy);
-      Value poolTensor = emptyZero(builder, poolTy, poolDynDims);
+  void getConv2DOpIndices(bool isNHWC, int &filterFDimIndex,
+                          int &filterCDimIndex, int &filterHDimIndex,
+                          int &filterWDimIndex, int &inputNDimIndex,
+                          int &inputHDimIndex, int &inputWDimIndex,
+                          int &inputCDimIndex) const {
```

Comment:
Have you tried `inferConvolutionDims` method? It should return the dimensions for you.

https://github.com/llvm/llvm-project/blob/02403f4e450b86d93197dd34045ff40a34b21494/mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp#L817-L818

Side note: we do not prefer passing many "output" arguments by reference. Instead, we can declare a struct and return the struct at the end. It generally improves the readability and reduces the bugs. In some cases reference parameters can bind to temporaries, leading to lifetime bugs. Having a struct also makes it more robust because people do not need to check the order carefully. They can access the member directly.

See https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs and https://abseil.io/tips/176 for mor details.

Reply:
Thanks for the pointers! @Max191 mentioned the same for `inferConvolutionDims` in another context. I spent a while looking for this but ended not finding it and writing my own hard coded version for grabbing conv dimensions :-p

The only question remains would be whether this can work with quantized conv types like `Conv2DNhwcHwcfQOp`. I'll give it a shot once I'm done with another high priority task and get back to this review. In the mean time, I'm marking this PR as draft in case it get accidentally merged.